## BACKGROUND AND BUSINESS PROBLEM
 
Since mid-2016 to January 31st 2020, the British government was dealing with Brexit and the British withdrawal from the European Union had affected the housing market for both citizens and overseas buyers as the government made it more expensive for landlords and overseas buyers to purchase houses (https://www.bullionvault.com/gold-news/uk-house-prices-022420201). Especially, real estate in London comes at quite a steep price. It is this exact reason why you should not settle and find only the best fit for your needs (https://data.london.gov.uk/housing/housing-market-report/). Therefore, I am going to use the latest data for buyers to make wise and effective decisions.

BUSINESS PROBLEM: What is the wisest investment option for people who seek a house in London now that Brexit is finalised?

SOLUTION: In order to recommend venues and the current average price of real estate, the neighbourhoods of London will be clustered. Recommendations of profitable venues according to amenities and essential facilities such as schools, hospitals & supermarkets will be provided.

## Data

Data on London properties and the relative price paid data were extracted from the HM Land Registry (http://landregistry.data.gov.uk/). If the building is divided into flats, there will be a SAON; Street; Locality; Town/City; District; County.

For data visialisation, FourSquare API interface will be used. Then the collected data from HM Land Registry and Foursquare API will be merged and show the most profitable investments in London.

## Methodology

The Methodology for this project will start with 1. Collection of Data
then continue with 2. Exploring and Understanding Data, 3. Data Preparation and Preprocessing 
and 4. Modelling

## 1. Collection of Data

I will start with importing the necessary libraries, then download the data from the HM Land Registry website as follows:

In [1]:
import os # Operating System
import numpy as np
import pandas as pd
import datetime as dt # Datetime
import json # library to handle JSON files

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

!conda install -c conda-forge folium=0.5.0 --yes
import folium #import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /Users/zeynepbokeer/anaconda3

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1g             |       h0b31af3_1         1.9 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         1.9 MB

The following packages will be UPDATED:

  openssl                                 1.1.1g-h0b31af3_0 --> 1.1.1g-h0b31af3_1



openssl-1.1.1g       | 1.9 MB    | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Solving environment: done

# All requested packages already installed.

Libraries imported.


In [6]:
#Reading the data
df_ppd = pd.read_csv("http://prod2.publicdata.landregistry.gov.uk.s3-website-eu-west-1.amazonaws.com/pp-2018.csv")

## 2. Explore and Understand Data

I read the dataset that I collected from the HM Land Registry website into a pandas' data frame and checking the raw data by running 'head'

In [9]:
df_ppd.head()

{7011B109-CFCA-8ED6-E053-6B04A8C075C1}  280000  2018-06-04 00:00  IP4 5ES  \
0  {7011B109-CFCB-8ED6-E053-6B04A8C075C1}  280000  2018-05-29 00:00  IP1 4BS   
1  {7011B109-CFCC-8ED6-E053-6B04A8C075C1}  170000  2018-04-27 00:00  IP4 4BH   
2  {7011B109-CFCD-8ED6-E053-6B04A8C075C1}  246000  2018-05-25 00:00  IP1 6NB   
3  {7011B109-CFCE-8ED6-E053-6B04A8C075C1}  180000  2018-06-08 00:00  IP3 9LZ   
4  {7011B109-CFCF-8ED6-E053-6B04A8C075C1}  245000  2018-05-11 00:00  IP1 4BU   

   S  N  F    3 Unnamed: 8   RANDWELL CLOSE Unnamed: 10  IPSWICH IPSWICH.1  \
0  T  N  F  261        NaN     NORWICH ROAD         NaN  IPSWICH   IPSWICH   
1  T  N  F   31        NaN      PARADE ROAD         NaN  IPSWICH   IPSWICH   
2  S  N  F   42        NaN    ELMCROFT ROAD         NaN  IPSWICH   IPSWICH   
3  T  N  F   48        NaN  WYNTERTON CLOSE         NaN  IPSWICH   IPSWICH   
4  T  N  F  235        NaN     NORWICH ROAD         NaN  IPSWICH   IPSWICH   

   SUFFOLK  A A.1  
0  SUFFOLK  A   A  
1  SUFFOLK  A   A  
2  SUFFOLK  A   A  
3  SUFFOLK  A   A  
4  SUFFOLK  A   A

Checking the number of rows and columns

In [10]:
df_ppd.shape

(1030277, 16)

As you can see, we have 1030277 columns and 16 rows

## 3. Data preparation and preprocessing

To prepare the dataset for the modeling process, I have to perform below-mentioned steps:

1- Rename the column names

2- Format the date column

3- Sort data by date of sale

4- Select data only for the city of London

5- Make a list of street names in London

6- Calculate the street-wise average price of the property

7- Read the street-wise coordinates into a data frame, eliminating recurring word London from individual names

8- Join the data to find the coordinates of locations which fit into my budget

9- Plot recommended locations on London map along with current market prices

In [11]:
# Rename the column names
df_ppd.columns = ['TUID', 'Price', 'Date_Transfer', 'Postcode', 'Prop_Type', 'Old_New', 'Duration', 'PAON', \
                  'SAON', 'Street', 'Locality', 'Town_City', 'District', 'County', 'PPD_Cat_Type', 'Record_Status']

In [12]:
# Format the date column
df_ppd['Date_Transfer'] = df_ppd['Date_Transfer'].apply(pd.to_datetime)

# Delete all obsolete transactions which were done before 2017
df_ppd.drop(df_ppd[df_ppd.Date_Transfer.dt.year < 2017].index, inplace=True)

# Sort by Date of Sale
df_ppd.sort_values(by=['Date_Transfer'],ascending=[False],inplace=True)


In [13]:
df_ppd_london = df_ppd.query("Town_City == 'LONDON'")

# Make a list of street names in London
streets = df_ppd_london['Street'].unique().tolist()

In [14]:
df_grp_price = df_ppd_london.groupby(['Street'])['Price'].mean().reset_index()

# Give meaningful names to the columns
df_grp_price.columns = ['Street', 'Avg_Price']

In [15]:
# Assign an Upper and Lower Limit to my budget
df_affordable = df_grp_price.query("(Avg_Price >= 2000000) & (Avg_Price <= 2200000)")

In [16]:
# Display the dataframe
df_affordable

Street     Avg_Price
146          AIREDALE AVENUE  2.022500e+06
197            ALBION STREET  2.096667e+06
412              ANTHONY WAY  2.000000e+06
554    ASHCHURCH PARK VILLAS  2.150000e+06
673              AVENUE ROAD  2.143471e+06
762          BALLINGDON ROAD  2.105000e+06
1092        BELSIZE CRESCENT  2.000000e+06
1135       BERESFORD TERRACE  2.100000e+06
1193          BETTRIDGE ROAD  2.025000e+06
1384            BLYTHS WHARF  2.000000e+06
1435            BOSTON PLACE  2.167500e+06
1515     BRACKENBURY GARDENS  2.150000e+06
1559         BRAMSHOT AVENUE  2.177900e+06
1704         BROADLANDS ROAD  2.043000e+06
1751        BRONDESBURY PARK  2.090357e+06
1775         BROOKFIELD PARK  2.150000e+06
1823          BROWNING CLOSE  2.160000e+06
1867        BRYANSTON SQUARE  2.197583e+06
1911            BUNKERS HILL  2.000000e+06
2019          BYWATER STREET  2.000000e+06
2072           CALTON AVENUE  2.016667e+06
2153        CANFIELD GARDENS  2.188333e+06
2226           CARLISLE ROAD  2.200000e+06
2242           CARLYLE CLOSE  2.175000e+06
2697      CHOLMELEY CRESCENT  2.025000e+06
2789       CLARE LAWN AVENUE  2.083000e+06
2857        CLEVELAND SQUARE  2.150000e+06
2989          COLINETTE ROAD  2.124375e+06
2997           COLLEGE CROSS  2.185000e+06
3042          COLVILLE PLACE  2.095000e+06
...                      ...           ...
9960          PLAYHOUSE YARD  2.172000e+06
10019            PONT STREET  2.070000e+06
10194       PROVIDENCE PLACE  2.063750e+06
10226      PUTNEY HEATH LANE  2.150000e+06
10275    QUEENS GATE GARDENS  2.181466e+06
10284      QUEENSBERRY PLACE  2.024667e+06
10314            RACTON ROAD  2.125000e+06
10402             RAVEN ROAD  2.020000e+06
10824          ROTHBURY ROAD  2.022100e+06
11263         SHELDON SQUARE  2.100000e+06
11425         SLOANE GARDENS  2.075000e+06
11500     SOUTH LAMBETH ROAD  2.029255e+06
11746      ST JAMES'S STREET  2.077857e+06
11856      ST QUINTIN AVENUE  2.015400e+06
11880         STADIUM STREET  2.050000e+06
11884        STAFFORD STREET  2.150000e+06
12141           SULIVAN ROAD  2.006667e+06
12323          TANNER STREET  2.023350e+06
12642           TIERNEY LANE  2.120500e+06
12664            TITE STREET  2.119807e+06
12927           TYSOE STREET  2.067500e+06
13205               WALBROOK  2.000000e+06
13312          WARWICK COURT  2.017500e+06
13405        WEDGWOOD VILLAS  2.000000e+06
13514      WEST TEMPLE SHEEN  2.012500e+06
13717             WILBY MEWS  2.017500e+06
13819        WINDMILL STREET  2.006667e+06
13865       WINTERBROOK ROAD  2.100000e+06
13920       WOODBOROUGH ROAD  2.125000e+06
14057            WREN STREET  2.050000e+06

[110 rows x 2 columns]

In [47]:
import pandas as pd
import numpy as np
import datetime as DT
import hmac
from geopy.geocoders import Nominatim
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [48]:
for index, item in df_affordable.iterrows():
    print(f"index: {index}")
    print(f"item: {item}")
    print(f"item.Street only: {item.Street}")

index: 146
item: Street                AIREDALE AVENUE
Avg_Price                  2.0225e+06
city_coord    (53.4363678, -1.106305)
Name: 146, dtype: object
item.Street only: AIREDALE AVENUE
index: 197
item: Street                     ALBION STREET
Avg_Price                    2.09667e+06
city_coord    (-33.8197594, 151.0102727)
Name: 197, dtype: object
item.Street only: ALBION STREET
index: 412
item: Street                     ANTHONY WAY
Avg_Price                        2e+06
city_coord    (44.094651, -123.151242)
Name: 412, dtype: object
item.Street only: ANTHONY WAY
index: 554
item: Street           ASHCHURCH PARK VILLAS
Avg_Price                     2.15e+06
city_coord    (51.5000507, -0.2421733)
Name: 554, dtype: object
item.Street only: ASHCHURCH PARK VILLAS
index: 673
item: Street                    AVENUE ROAD
Avg_Price                 2.14347e+06
city_coord    (51.4067969, -0.049519)
Name: 673, dtype: object
item.Street only: AVENUE ROAD
index: 762
item: Street                

In [49]:
geolocator = Nominatim()

ConfigurationError: Using Nominatim with default or sample `user_agent` "geopy/2.0.0" is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`.

In [50]:
df_affordable['city_coord'] = df_affordable['Street'].apply(geolocator.geocode).apply(lambda x: (x.latitude, x.longitude))

/Users/zeynepbokeer/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [51]:
df_affordable

Street     Avg_Price  \
146          AIREDALE AVENUE  2.022500e+06   
197            ALBION STREET  2.096667e+06   
412              ANTHONY WAY  2.000000e+06   
554    ASHCHURCH PARK VILLAS  2.150000e+06   
673              AVENUE ROAD  2.143471e+06   
762          BALLINGDON ROAD  2.105000e+06   
1092        BELSIZE CRESCENT  2.000000e+06   
1135       BERESFORD TERRACE  2.100000e+06   
1193          BETTRIDGE ROAD  2.025000e+06   
1384            BLYTHS WHARF  2.000000e+06   
1435            BOSTON PLACE  2.167500e+06   
1515     BRACKENBURY GARDENS  2.150000e+06   
1559         BRAMSHOT AVENUE  2.177900e+06   
1704         BROADLANDS ROAD  2.043000e+06   
1751        BRONDESBURY PARK  2.090357e+06   
1775         BROOKFIELD PARK  2.150000e+06   
1823          BROWNING CLOSE  2.160000e+06   
1867        BRYANSTON SQUARE  2.197583e+06   
1911            BUNKERS HILL  2.000000e+06   
2019          BYWATER STREET  2.000000e+06   
2072           CALTON AVENUE  2.016667e+06   
2153        CANFIELD GARDENS  2.188333e+06   
2226           CARLISLE ROAD  2.200000e+06   
2242           CARLYLE CLOSE  2.175000e+06   
2697      CHOLMELEY CRESCENT  2.025000e+06   
2789       CLARE LAWN AVENUE  2.083000e+06   
2857        CLEVELAND SQUARE  2.150000e+06   
2989          COLINETTE ROAD  2.124375e+06   
2997           COLLEGE CROSS  2.185000e+06   
3042          COLVILLE PLACE  2.095000e+06   
...                      ...           ...   
9960          PLAYHOUSE YARD  2.172000e+06   
10019            PONT STREET  2.070000e+06   
10194       PROVIDENCE PLACE  2.063750e+06   
10226      PUTNEY HEATH LANE  2.150000e+06   
10275    QUEENS GATE GARDENS  2.181466e+06   
10284      QUEENSBERRY PLACE  2.024667e+06   
10314            RACTON ROAD  2.125000e+06   
10402             RAVEN ROAD  2.020000e+06   
10824          ROTHBURY ROAD  2.022100e+06   
11263         SHELDON SQUARE  2.100000e+06   
11425         SLOANE GARDENS  2.075000e+06   
11500     SOUTH LAMBETH ROAD  2.029255e+06   
11746      ST JAMES'S STREET  2.077857e+06   
11856      ST QUINTIN AVENUE  2.015400e+06   
11880         STADIUM STREET  2.050000e+06   
11884        STAFFORD STREET  2.150000e+06   
12141           SULIVAN ROAD  2.006667e+06   
12323          TANNER STREET  2.023350e+06   
12642           TIERNEY LANE  2.120500e+06   
12664            TITE STREET  2.119807e+06   
12927           TYSOE STREET  2.067500e+06   
13205               WALBROOK  2.000000e+06   
13312          WARWICK COURT  2.017500e+06   
13405        WEDGWOOD VILLAS  2.000000e+06   
13514      WEST TEMPLE SHEEN  2.012500e+06   
13717             WILBY MEWS  2.017500e+06   
13819        WINDMILL STREET  2.006667e+06   
13865       WINTERBROOK ROAD  2.100000e+06   
13920       WOODBOROUGH ROAD  2.125000e+06   
14057            WREN STREET  2.050000e+06   

                                     city_coord  
146                     (53.4363678, -1.106305)  
197                  (-33.8197594, 151.0102727)  
412                    (44.094651, -123.151242)  
554                    (51.5000507, -0.2421733)  
673                     (51.4067969, -0.049519)  
762                    (51.4541892, -0.1588555)  
1092                   (51.5495675, -0.1727325)  
1135                   (55.9757427, -3.2031975)  
1193                   (57.0320376, -2.1476985)  
1384       (51.50900105, -0.034583431747228985)  
1435         (51.5073726, -0.34033780731250823)  
1515                    (51.5006233, -0.230729)  
1559                  (44.6060675, -75.6904856)  
1704                    (52.7501148, 1.5662465)  
1751                   (51.5405138, -0.2103359)  
1775                  (36.9315361, -76.2327167)  
1823                    (52.6640289, 1.2531443)  
1867                    (51.517067, -0.1603646)  
1911                   (53.0180584, -2.3802677)  
2019                    (42.589268, -83.443243)  
2072                    (53.5087599, -2.283024)  
2153                    (51.546792, -0.1795496)  
2226               

In [52]:
df_affordable[['Latitude', 'Longitude']] = df_affordable['city_coord'].apply(pd.Series)

/Users/zeynepbokeer/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:3391: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[k1] = value[k2]


In [53]:
df_affordable

Street     Avg_Price  \
146          AIREDALE AVENUE  2.022500e+06   
197            ALBION STREET  2.096667e+06   
412              ANTHONY WAY  2.000000e+06   
554    ASHCHURCH PARK VILLAS  2.150000e+06   
673              AVENUE ROAD  2.143471e+06   
762          BALLINGDON ROAD  2.105000e+06   
1092        BELSIZE CRESCENT  2.000000e+06   
1135       BERESFORD TERRACE  2.100000e+06   
1193          BETTRIDGE ROAD  2.025000e+06   
1384            BLYTHS WHARF  2.000000e+06   
1435            BOSTON PLACE  2.167500e+06   
1515     BRACKENBURY GARDENS  2.150000e+06   
1559         BRAMSHOT AVENUE  2.177900e+06   
1704         BROADLANDS ROAD  2.043000e+06   
1751        BRONDESBURY PARK  2.090357e+06   
1775         BROOKFIELD PARK  2.150000e+06   
1823          BROWNING CLOSE  2.160000e+06   
1867        BRYANSTON SQUARE  2.197583e+06   
1911            BUNKERS HILL  2.000000e+06   
2019          BYWATER STREET  2.000000e+06   
2072           CALTON AVENUE  2.016667e+06   
2153        CANFIELD GARDENS  2.188333e+06   
2226           CARLISLE ROAD  2.200000e+06   
2242           CARLYLE CLOSE  2.175000e+06   
2697      CHOLMELEY CRESCENT  2.025000e+06   
2789       CLARE LAWN AVENUE  2.083000e+06   
2857        CLEVELAND SQUARE  2.150000e+06   
2989          COLINETTE ROAD  2.124375e+06   
2997           COLLEGE CROSS  2.185000e+06   
3042          COLVILLE PLACE  2.095000e+06   
...                      ...           ...   
9960          PLAYHOUSE YARD  2.172000e+06   
10019            PONT STREET  2.070000e+06   
10194       PROVIDENCE PLACE  2.063750e+06   
10226      PUTNEY HEATH LANE  2.150000e+06   
10275    QUEENS GATE GARDENS  2.181466e+06   
10284      QUEENSBERRY PLACE  2.024667e+06   
10314            RACTON ROAD  2.125000e+06   
10402             RAVEN ROAD  2.020000e+06   
10824          ROTHBURY ROAD  2.022100e+06   
11263         SHELDON SQUARE  2.100000e+06   
11425         SLOANE GARDENS  2.075000e+06   
11500     SOUTH LAMBETH ROAD  2.029255e+06   
11746      ST JAMES'S STREET  2.077857e+06   
11856      ST QUINTIN AVENUE  2.015400e+06   
11880         STADIUM STREET  2.050000e+06   
11884        STAFFORD STREET  2.150000e+06   
12141           SULIVAN ROAD  2.006667e+06   
12323          TANNER STREET  2.023350e+06   
12642           TIERNEY LANE  2.120500e+06   
12664            TITE STREET  2.119807e+06   
12927           TYSOE STREET  2.067500e+06   
13205               WALBROOK  2.000000e+06   
13312          WARWICK COURT  2.017500e+06   
13405        WEDGWOOD VILLAS  2.000000e+06   
13514      WEST TEMPLE SHEEN  2.012500e+06   
13717             WILBY MEWS  2.017500e+06   
13819        WINDMILL STREET  2.006667e+06   
13865       WINTERBROOK ROAD  2.100000e+06   
13920       WOODBOROUGH ROAD  2.125000e+06   
14057            WREN STREET  2.050000e+06   

                                     city_coord   Latitude   Longitude  
146                     (53.4363678, -1.106305)  53.436368   -1.106305  
197                  (-33.8197594, 151.0102727) -33.819759  151.010273  
412                    (44.094651, -123.151242)  44.094651 -123.151242  
554                    (51.5000507, -0.2421733)  51.500051   -0.242173  
673                     (51.4067969, -0.049519)  51.406797   -0.049519  
762                    (51.4541892, -0.1588555)  51.454189   -0.158856  
1092                   (51.5495675, -0.1727325)  51.549568   -0.172733  
1135                   (55.9757427, -3.2031975)  55.975743   -3.203197  
1193                   (57.0320376, -2.1476985)  57.032038   -2.147699  
1384       (51.50900105, -0.034583431747228985)  51.509001   -0.034583  
1435         (51.5073726, -0.34033780731250823)  51.507373   -0.340338  
1515                    (51.5006233, -0.230729)  51.500623   -0.230729  
1559                  (44.6060675, -75.6904856)  44.606068  -75.690486  
1704                    (52.7501148, 1.5662465)  52.750115    1.566247  
1751                   (51.5405138, -0.2103359)  51.540514   -0.210336  
1

In [54]:
df = df_affordable.drop(columns=['city_coord'])

In [55]:
df

Street     Avg_Price   Latitude   Longitude
146          AIREDALE AVENUE  2.022500e+06  53.436368   -1.106305
197            ALBION STREET  2.096667e+06 -33.819759  151.010273
412              ANTHONY WAY  2.000000e+06  44.094651 -123.151242
554    ASHCHURCH PARK VILLAS  2.150000e+06  51.500051   -0.242173
673              AVENUE ROAD  2.143471e+06  51.406797   -0.049519
762          BALLINGDON ROAD  2.105000e+06  51.454189   -0.158856
1092        BELSIZE CRESCENT  2.000000e+06  51.549568   -0.172733
1135       BERESFORD TERRACE  2.100000e+06  55.975743   -3.203197
1193          BETTRIDGE ROAD  2.025000e+06  57.032038   -2.147699
1384            BLYTHS WHARF  2.000000e+06  51.509001   -0.034583
1435            BOSTON PLACE  2.167500e+06  51.507373   -0.340338
1515     BRACKENBURY GARDENS  2.150000e+06  51.500623   -0.230729
1559         BRAMSHOT AVENUE  2.177900e+06  44.606068  -75.690486
1704         BROADLANDS ROAD  2.043000e+06  52.750115    1.566247
1751        BRONDESBURY PARK  2.090357e+06  51.540514   -0.210336
1775         BROOKFIELD PARK  2.150000e+06  36.931536  -76.232717
1823          BROWNING CLOSE  2.160000e+06  52.664029    1.253144
1867        BRYANSTON SQUARE  2.197583e+06  51.517067   -0.160365
1911            BUNKERS HILL  2.000000e+06  53.018058   -2.380268
2019          BYWATER STREET  2.000000e+06  42.589268  -83.443243
2072           CALTON AVENUE  2.016667e+06  53.508760   -2.283024
2153        CANFIELD GARDENS  2.188333e+06  51.546792   -0.179550
2226           CARLISLE ROAD  2.200000e+06  43.399438  -79.973215
2242           CARLYLE CLOSE  2.175000e+06  52.085813    1.135452
2697      CHOLMELEY CRESCENT  2.025000e+06  51.572397   -0.145105
2789       CLARE LAWN AVENUE  2.083000e+06  51.457853   -0.262927
2857        CLEVELAND SQUARE  2.150000e+06  53.401632   -2.985580
2989          COLINETTE ROAD  2.124375e+06  51.461939   -0.229873
2997           COLLEGE CROSS  2.185000e+06  51.949009   -8.549684
3042          COLVILLE PLACE  2.095000e+06  57.155024   -2.092671
...                      ...           ...        ...         ...
9960          PLAYHOUSE YARD  2.172000e+06  51.512740   -0.102746
10019            PONT STREET  2.070000e+06  50.780607    6.079629
10194       PROVIDENCE PLACE  2.063750e+06  29.677621  -98.123832
10226      PUTNEY HEATH LANE  2.150000e+06  51.452337   -0.221080
10275    QUEENS GATE GARDENS  2.181466e+06  51.407490    0.077369
10284      QUEENSBERRY PLACE  2.024667e+06  51.547738    0.047427
10314            RACTON ROAD  2.125000e+06  50.861234   -0.935288
10402             RAVEN ROAD  2.020000e+06  51.611329   -1.271216
10824          ROTHBURY ROAD  2.022100e+06 -41.863688  147.231291
11263         SHELDON SQUARE  2.100000e+06  51.519392   -0.180058
11425         SLOANE GARDENS  2.075000e+06  51.368675    0.069859
11500     SOUTH LAMBETH ROAD  2.029255e+06  51.480064   -0.124200
11746      ST JAMES'S STREET  2.077857e+06  51.441431    0.363312
11856      ST QUINTIN AVENUE  2.015400e+06  51.518869   -0.217381
11880         STADIUM STREET  2.050000e+06  41.321684  -75.810943
11884        STAFFORD STREET  2.150000e+06 -45.883947  170.490855
12141           SULIVAN ROAD  2.006667e+06  51.466617   -0.196085
12323          TANNER STREET  2.023350e+06  48.330030   12.981914
12642           TIERNEY LANE  2.120500e+06  41.838381  -88.290844
12664            TITE STREET  2.119807e+06  51.485949   -0.161080
12927           TYSOE STREET  2.067500e+06  51.526513   -0.108234
13205               WALBROOK  2.000000e+06  39.311736  -76.668471
13312          WARWICK COURT  2.017500e+06  42.707025  -71.190460
13405        WEDGWOOD VILLAS  2.000000e+06  30.331655  -81.470446
13514      WEST TEMPLE SHEEN  2.012500e+06  51.460139   -0.275404
13717             WILBY MEWS  2.017500e+06  51.509335   -0.202950
13819        WINDMILL STREET  2.006667e+06  29.785269  -98.455945
13865       WINTERBROOK ROAD  2.100000e+06  39.372628  -76.785686
13920       WOODBOROUGH ROAD  2.125000e+06  52.959589   -1.149

In [57]:
address = 'London, UK'

location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of London City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of London City are 51.5073219, -0.1276474.


In [58]:
# create map of London using latitude and longitude values
map_london = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, price, street in zip(df['Latitude'], df['Longitude'], df['Avg_Price'], df['Street']):
    label = '{}, {}'.format(street, price)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_london)  
    
map_london

In [62]:
#Define Foursquare Credentials and Version

CLIENT_ID = 'KRTD1FJV3FCEIKJELRVOU5V4IMO1RQPFO0NK0GYZK52GBELH' # Foursquare ID
CLIENT_SECRET = 'JFFWFIJDN0RG2DHGS1G0UQH44WRDMZXBJHIWOAPI2F4EWGHK' # Foursquare Secret
VERSION = '20200730' # Foursquare API version

print('Your credentials:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentials:
CLIENT_ID: KRTD1FJV3FCEIKJELRVOU5V4IMO1RQPFO0NK0GYZK52GBELH
CLIENT_SECRET:JFFWFIJDN0RG2DHGS1G0UQH44WRDMZXBJHIWOAPI2F4EWGHK


Now that everything is ready, I can proceed to the Modeling phase. I will analyse neighbourhoods so that I can recommend the most profitable investments for their needs. As mentioned in the Background Section, recommendations of profitable venues according to amenities and essential facilities such as schools, hospitals & supermarkets will be provided.

## 4. Modeling
I will use the k-means clustering technique will be used for this project as it is accurate, efficient and highly flexible.

In [63]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Street', 
                  'Street Latitude', 
                  'Street Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [64]:
# Run the above function on each location and create a new dataframe called location_venues and display it.
location_venues = getNearbyVenues(names=df['Street'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

AIREDALE AVENUE
ALBION STREET
ANTHONY WAY
ASHCHURCH PARK VILLAS
AVENUE ROAD
BALLINGDON ROAD
BELSIZE CRESCENT
BERESFORD TERRACE
BETTRIDGE ROAD
BLYTHS WHARF
BOSTON PLACE
BRACKENBURY GARDENS
BRAMSHOT AVENUE
BROADLANDS ROAD
BRONDESBURY PARK
BROOKFIELD PARK
BROWNING CLOSE
BRYANSTON SQUARE
BUNKERS HILL
BYWATER STREET
CALTON AVENUE
CANFIELD GARDENS
CARLISLE ROAD
CARLYLE CLOSE
CHOLMELEY CRESCENT
CLARE LAWN AVENUE
CLEVELAND SQUARE
COLINETTE ROAD
COLLEGE CROSS
COLVILLE PLACE
COTSWOLD MEWS
COVERDALE ROAD
CRANLEY MEWS
CUMBERLAND TERRACE
DENBIGH CLOSE
DONNE PLACE
DRYBURGH ROAD
DUNSTAN ROAD
ECCLESTON MEWS
EGERTON PLACE
ELSWORTHY ROAD
ESSEX PARK MEWS
EYNELLA ROAD
GLOUCESTER SQUARE
GRANARD ROAD
GREAT RUSSELL STREET
GREEN CLOSE
GROSVENOR GARDENS
GUION ROAD
HAMBLEDON PLACE
HANOVER YARD
HENDERSON ROAD
HENDON WOOD LANE
HEWER STREET
HIGHLEVER ROAD
HILLGATE PLACE
HOBHOUSE COURT
HOLLAND PARK
IVOR PLACE
KINGS GATE WALK
LAURIER ROAD
LEINSTER MEWS
LONG LANE
MANSON MEWS
MARGIN DRIVE
MAUNSEL STREET
MEADOWBANK
MEL

In [65]:
location_venues

Street  Street Latitude  Street Longitude  \
0           AIREDALE AVENUE        53.436368         -1.106305   
1           AIREDALE AVENUE        53.436368         -1.106305   
2           AIREDALE AVENUE        53.436368         -1.106305   
3           AIREDALE AVENUE        53.436368         -1.106305   
4             ALBION STREET       -33.819759        151.010273   
5             ALBION STREET       -33.819759        151.010273   
6             ALBION STREET       -33.819759        151.010273   
7             ALBION STREET       -33.819759        151.010273   
8             ALBION STREET       -33.819759        151.010273   
9             ALBION STREET       -33.819759        151.010273   
10            ALBION STREET       -33.819759        151.010273   
11            ALBION STREET       -33.819759        151.010273   
12            ALBION STREET       -33.819759        151.010273   
13            ALBION STREET       -33.819759        151.010273   
14            ALBION STREET       -33.819759        151.010273   
15            ALBION STREET       -33.819759        151.010273   
16    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
17    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
18    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
19    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
20    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
21    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
22    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
23    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
24    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
25    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
26    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
27    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
28    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
29    ASHCHURCH PARK VILLAS        51.500051         -0.242173   
...                     ...              ...               ...   
2495       WOODBOROUGH ROAD        52.959589         -1.149200   
2496       WOODBOROUGH ROAD        52.959589         -1.149200   
2497       WOODBOROUGH ROAD        52.959589         -1.149200   
2498       WOODBOROUGH ROAD        52.959589         -1.149200   
2499       WOODBOROUGH ROAD        52.959589         -1.149200   
2500       WOODBOROUGH ROAD        52.959589         -1.149200   
2501       WOODBOROUGH ROAD        52.959589         -1.149200   
2502       WOODBOROUGH ROAD        52.959589         -1.149200   
2503       WOODBOROUGH ROAD        52.959589         -1.149200   
2504       WOODBOROUGH ROAD        52.959589         -1.149200   
2505       WOODBOROUGH ROAD        52.959589         -1.149200   
2506       WOODBOROUGH ROAD        52.959589         -1.149200   
2507       WOODBOROUGH ROAD        52.959589         -1.149200   
2508       WOODBOROUGH ROAD        52.959589         -1.149200   
2509       WOODBOROUGH ROAD        52.959589         -1.149200   
2510       WOODBOROUGH ROAD        52.959589         -1.149200   
2511       WOODBOROUGH ROAD        52.959589         -1.149200   
2512       WOODBOROUGH ROAD        52.959589         -1.149200   
2513       WOODBOROUGH ROAD        52.959589         -1.149200   
2514       WOODBOROUGH ROAD        52.959589         -1.149200   
2515       WOODBOROUGH ROAD        52.959589         -1.149200   
2516       WOODBOROUGH ROAD        52.959589         -1.149200   
2517       WOODBOROUGH ROAD        52.959589         -1.149200   
2518       WOODBOROUGH ROAD        52.959589         -1.149200   
2519       WOODBOROUGH ROAD        52.959589         -1.149200   
2520       WOODBOROUGH ROAD        52.959589         -1.149200   
2521       WOODBOROUGH ROAD        52.959589         -1.149200   
2522       WOODBOROUGH ROAD        52.959589         -1.149200   
2523            WREN STREET        42.621784        -88.27508

In [66]:
location_venues.groupby('Street').count()

Street Latitude  Street Longitude  Venue  \
Street                                                            
AIREDALE AVENUE                      4                 4      4   
ALBION STREET                       12                12     12   
ASHCHURCH PARK VILLAS               29                29     29   
AVENUE ROAD                          4                 4      4   
BALLINGDON ROAD                     10                10     10   
BELSIZE CRESCENT                    13                13     13   
BERESFORD TERRACE                    5                 5      5   
BETTRIDGE ROAD                       5                 5      5   
BLYTHS WHARF                        17                17     17   
BOSTON PLACE                        14                14     14   
BRACKENBURY GARDENS                 16                16     16   
BRAMSHOT AVENUE                      4                 4      4   
BROADLANDS ROAD                      4                 4      4   
BRONDESBURY PARK                     8                 8      8   
BROOKFIELD PARK                      3                 3      3   
BROWNING CLOSE                       4                 4      4   
BRYANSTON SQUARE                   100               100    100   
BYWATER STREET                      26                26     26   
CALTON AVENUE                        4                 4      4   
CANFIELD GARDENS                    73                73     73   
CARLYLE CLOSE                        1                 1      1   
CHOLMELEY CRESCENT                  31                31     31   
CLARE LAWN AVENUE                    4                 4      4   
CLEVELAND SQUARE                    86                86     86   
COLINETTE ROAD                       9                 9      9   
COLLEGE CROSS                        4                 4      4   
COLVILLE PLACE                       4                 4      4   
COTSWOLD MEWS                       19                19     19   
CRANLEY MEWS                        56                56     56   
CUMBERLAND TERRACE                  20                20     20   
...                                ...               ...    ...   
PAULTONS SQUARE                     50                50     50   
PAVILION ROAD                        2                 2      2   
PENCOMBE MEWS                      100               100    100   
PERCIVAL ROAD                        5                 5      5   
PLAYHOUSE YARD                      70                70     70   
PONT STREET                         58                58     58   
PROVIDENCE PLACE                     3                 3      3   
PUTNEY HEATH LANE                    8                 8      8   
QUEENS GATE GARDENS                  7                 7      7   
QUEENSBERRY PLACE                    7                 7      7   
RACTON ROAD                          6                 6      6   
SHELDON SQUARE                      33                33     33   
SLOANE GARDENS                       2                 2      2   
SOUTH LAMBETH ROAD                  27                27     27   
ST JAMES'S STREET                   25                25     25   
ST QUINTIN AVENUE                    7                 7      7   
STADIUM STREET                      12                12     12   
STAFFORD STREET                      1                 1      1   
SULIVAN ROAD                         6                 6      6   
TANNER STREET                        1                 1      1   
TIERNEY LANE                         1                 1      1   
TITE STREET                         39                39     39   
TYSOE STREET                        91                91     91   
WALBROOK                             4                 4      4   
WARWICK COURT                        6                 6      6   
WEDGWOOD VILLAS                      3                 3      3   
WEST TEMPLE SHEEN                    5                 5      5   
WIL

In [67]:
# get the List of Unique Categories
print('There are {} uniques categories.'.format(len(location_venues['Venue Category'].unique())))

There are 289 uniques categories.


In [68]:
location_venues.shape

(2525, 7)

In [69]:
# one hot encoding
venues_onehot = pd.get_dummies(location_venues[['Venue Category']], prefix="", prefix_sep="")

# add street column back to dataframe
venues_onehot['Street'] = location_venues['Street'] 

# move street column to the first column
fixed_columns = [venues_onehot.columns[-1]] + list(venues_onehot.columns[:-1])

#fixed_columns
venues_onehot = venues_onehot[fixed_columns]

venues_onehot.head()

Street  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  AIREDALE AVENUE                  0               0                  0   
1  AIREDALE AVENUE                  0               0                  0   
2  AIREDALE AVENUE                  0               0                  0   
3  AIREDALE AVENUE                  0               0                  0   
4    ALBION STREET                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  \
0                   0                    0             1   
1                   0                    0             0   
2                   0                    0             0   
3                   0                    0             0   
4                   0                    0             0   

   Argentinian Restaurant  Art Gallery  Art Museum  ...  Video Game Store  \
0                       0            0           0  ...                 0   
1                       0            0           0  ...                 0   
2                       0            0           0  ...                 0   
3                       0            0           0  ...                 0   
4                       0            0           0  ...                 0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  \
0            0                      0                0           0         0   
1            0                      0                0           0         0   
2            0                      0                0           0         0   
3            0                      0                0           0         0   
4            0                      0                0           0         0   

   Wine Shop  Women's Store  Yoga Studio  Zoo  
0          0              0            0    0  
1          0              0            0    0  
2          0              0            0    0  
3          0              0            0    0  
4          0              0            0    0  

[5 rows x 290 columns]

In [70]:
GROUPED_LONDON = venues_onehot.groupby('Street').mean().reset_index()
GROUPED_LONDON

Street  Accessories Store  Adult Boutique  \
0         AIREDALE AVENUE               0.00             0.0   
1           ALBION STREET               0.00             0.0   
2   ASHCHURCH PARK VILLAS               0.00             0.0   
3             AVENUE ROAD               0.00             0.0   
4         BALLINGDON ROAD               0.00             0.0   
5        BELSIZE CRESCENT               0.00             0.0   
6       BERESFORD TERRACE               0.00             0.0   
7          BETTRIDGE ROAD               0.00             0.0   
8            BLYTHS WHARF               0.00             0.0   
9            BOSTON PLACE               0.00             0.0   
10    BRACKENBURY GARDENS               0.00             0.0   
11        BRAMSHOT AVENUE               0.00             0.0   
12        BROADLANDS ROAD               0.00             0.0   
13       BRONDESBURY PARK               0.00             0.0   
14        BROOKFIELD PARK               0.00             0.0   
15         BROWNING CLOSE               0.00             0.0   
16       BRYANSTON SQUARE               0.00             0.0   
17         BYWATER STREET               0.00             0.0   
18          CALTON AVENUE               0.00             0.0   
19       CANFIELD GARDENS               0.00             0.0   
20          CARLYLE CLOSE               0.00             0.0   
21     CHOLMELEY CRESCENT               0.00             0.0   
22      CLARE LAWN AVENUE               0.00             0.0   
23       CLEVELAND SQUARE               0.00             0.0   
24         COLINETTE ROAD               0.00             0.0   
25          COLLEGE CROSS               0.00             0.0   
26         COLVILLE PLACE               0.00             0.0   
27          COTSWOLD MEWS               0.00             0.0   
28           CRANLEY MEWS               0.00             0.0   
29     CUMBERLAND TERRACE               0.00             0.0   
..                    ...                ...             ...   
68        PAULTONS SQUARE               0.02             0.0   
69          PAVILION ROAD               0.00             0.0   
70          PENCOMBE MEWS               0.00             0.0   
71          PERCIVAL ROAD               0.00             0.0   
72         PLAYHOUSE YARD               0.00             0.0   
73            PONT STREET               0.00             0.0   
74       PROVIDENCE PLACE               0.00             0.0   
75      PUTNEY HEATH LANE               0.00             0.0   
76    QUEENS GATE GARDENS               0.00             0.0   
77      QUEENSBERRY PLACE               0.00             0.0   
78            RACTON ROAD               0.00             0.0   
79         SHELDON SQUARE               0.00             0.0   
80         SLOANE GARDENS               0.00             0.0   
81     SOUTH LAMBETH ROAD               0.00             0.0   
82      ST JAMES'S STREET               0.00             0.0   
83      ST QUINTIN AVENUE               0.00             0.0   
84         STADIUM STREET               0.00             0.0   
85        STAFFORD STREET               0.00             0.0   
86           SULIVAN ROAD               0.00             0.0   
87          TANNER STREET               0.00             0.0   
88           TIERNEY LANE               0.00             0.0   
89            TITE STREET               0.00             0.0   
90           TYSOE STREET               0.00             0.0   
91               WALBROOK               0.00             0.0   
92          WARWICK COURT               0.00             0.0   
93        WEDGWOOD VILLAS               0.00             0.0   
94      WEST TEMPLE SHEEN               0.00             0.0   
95             WILBY MEWS               0.00             0.0   
96       WOODBOROUGH ROAD               0.00             0.0   
97            WREN STREET               0.00             0.0   

    Afghan Restaurant  African Restaurant  America

In [71]:
GROUPED_LONDON.shape

(98, 290)

In [74]:
# What are the top 5 venues/facilities nearby profitable real estate investments?#

num_top_venues = 5

for hood in GROUPED_LONDON['Street']:
    print("----"+hood+"----")
    temp = GROUPED_LONDON[GROUPED_LONDON['Street'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----AIREDALE AVENUE----
                venue  freq
0                 Pub  0.25
1        Antique Shop  0.25
2            Pharmacy  0.25
3  Italian Restaurant  0.25
4   Accessories Store  0.00


----ALBION STREET----
               venue  freq
0  Indian Restaurant  0.33
1               Café  0.17
2         Sports Bar  0.08
3        Coffee Shop  0.08
4               Park  0.08


----ASHCHURCH PARK VILLAS----
               venue  freq
0      Grocery Store  0.14
1               Park  0.10
2                Pub  0.10
3  Indian Restaurant  0.07
4             Bakery  0.07


----AVENUE ROAD----
               venue  freq
0      Grocery Store  0.25
1   Tapas Restaurant  0.25
2               Park  0.25
3       Tram Station  0.25
4  Accessories Store  0.00


----BALLINGDON ROAD----
                venue  freq
0                 Pub   0.3
1       Women's Store   0.1
2           Wine Shop   0.1
3        Antique Shop   0.1
4  Italian Restaurant   0.1


----BELSIZE CRESCENT----
                venue  

                 venue  freq
0                  Pub  0.13
1          Coffee Shop  0.07
2    French Restaurant  0.05
3   Mexican Restaurant  0.03
4  Arts & Crafts Store  0.03


----HENDERSON ROAD----
                    venue  freq
0          Clothing Store  0.24
1  Furniture / Home Store  0.06
2          Ice Cream Shop  0.06
3              Shoe Store  0.06
4             Snack Place  0.06


----HENDON WOOD LANE----
                venue  freq
0  Athletics & Sports   1.0
1   Accessories Store   0.0
2        Noodle House   0.0
3     Organic Grocery   0.0
4        Optical Shop   0.0


----HEWER STREET----
                       venue  freq
0                       Café  0.14
1                Pizza Place  0.10
2                     Bakery  0.10
3                       Park  0.05
4  Middle Eastern Restaurant  0.05


----HIGHLEVER ROAD----
               venue  freq
0               Park   0.2
1       Soccer Field   0.2
2  Convenience Store   0.1
3                Pub   0.1
4      Grocery Store 

               venue  freq
0               Bank   1.0
1  Accessories Store   0.0
2    Organic Grocery   0.0
3       Optical Shop   0.0
4        Opera House   0.0


----TIERNEY LANE----
               venue  freq
0     Sandwich Place   1.0
1  Accessories Store   0.0
2  Other Repair Shop   0.0
3       Optical Shop   0.0
4        Opera House   0.0


----TITE STREET----
                  venue  freq
0                   Pub  0.08
1           Pizza Place  0.05
2  Gym / Fitness Center  0.05
3   Japanese Restaurant  0.05
4           Coffee Shop  0.05


----TYSOE STREET----
         venue  freq
0          Pub  0.18
1  Coffee Shop  0.07
2         Café  0.05
3         Park  0.05
4          Bar  0.04


----WALBROOK----
                 venue  freq
0    Convenience Store  0.25
1  American Restaurant  0.25
2          Gas Station  0.25
3                 Bank  0.25
4           Playground  0.00


----WARWICK COURT----
                venue  freq
0   Convenience Store  0.17
1         Pizza Place  0.17
2

In [75]:
# Define a function to return the most common venues/facilities nearby real estate investments#

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [76]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Street']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

In [78]:
# create a new dataframe
venues_sorted = pd.DataFrame(columns=columns)
venues_sorted['Street'] = GROUPED_LONDON['Street']

for ind in np.arange(GROUPED_LONDON.shape[0]):
    venues_sorted.iloc[ind, 1:] = return_most_common_venues(GROUPED_LONDON.iloc[ind, :], num_top_venues)

In [79]:
venues_sorted.head()

Street 1st Most Common Venue 2nd Most Common Venue  \
0        AIREDALE AVENUE    Italian Restaurant              Pharmacy   
1          ALBION STREET     Indian Restaurant                  Café   
2  ASHCHURCH PARK VILLAS         Grocery Store                  Park   
3            AVENUE ROAD      Tapas Restaurant         Grocery Store   
4        BALLINGDON ROAD                   Pub    Italian Restaurant   

  3rd Most Common Venue     4th Most Common Venue 5th Most Common Venue  \
0          Antique Shop                       Pub                   Zoo   
1                   Gym                Sports Bar           Coffee Shop   
2                   Pub  Mediterranean Restaurant           Coffee Shop   
3                  Park              Tram Station                   Zoo   
4         Women's Store                 Wine Shop                Bakery   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0     Electronics Store    English Restaurant  Ethiopian Restaurant   
1          Dessert Shop                   Pub                  Park   
2     Fish & Chips Shop     Indian Restaurant                Bakery   
3            Donut Shop     Electronics Store    English Restaurant   
4           Coffee Shop          Antique Shop                  Café   

  9th Most Common Venue 10th Most Common Venue  
0           Event Space                Exhibit  
1                   Zoo     English Restaurant  
2   Japanese Restaurant                   Café  
3  Ethiopian Restaurant            Event Space  
4                   Zoo     English Restaurant

In [80]:
venues_sorted.shape

(98, 11)

In [81]:
GROUPED_LONDON.shape

(98, 290)

In [82]:
GROUPED_LONDON=df

Now that we can see the most profitable real estate investments in London, we can cluster properties by venues/facilities/amenities nearby.

In [85]:
#Distribute in 5 Clusters

# set number of clusters
kclusters = 5

GROUPED_LONDON_clustering = GROUPED_LONDON.drop('Street', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(GROUPED_LONDON_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:50]

array([0, 2, 0, 1, 1, 2, 0, 2, 0, 0, 1, 1, 3, 4, 2, 1, 1, 3, 0, 0, 0, 3,
       3, 3, 0, 4, 1, 2, 3, 2, 1, 4, 1, 2, 4, 0, 1, 0, 2, 3, 4, 2, 0, 4,
       0, 1, 2, 2, 4, 2], dtype=int32)

In [110]:
#Dataframe to include Clusters

GROUPED_LONDON_clustering=df
GROUPED_LONDON_clustering.head()

Street     Avg_Price   Latitude   Longitude
146        AIREDALE AVENUE  2.022500e+06  53.436368   -1.106305
197          ALBION STREET  2.096667e+06 -33.819759  151.010273
412            ANTHONY WAY  2.000000e+06  44.094651 -123.151242
554  ASHCHURCH PARK VILLAS  2.150000e+06  51.500051   -0.242173
673            AVENUE ROAD  2.143471e+06  51.406797   -0.049519

In [111]:
GROUPED_LONDON_clustering.shape

(110, 4)

In [112]:
df.shape

(110, 4)

In [113]:
GROUPED_LONDON_clustering.dtypes

Street        object
Avg_Price    float64
Latitude     float64
Longitude    float64
dtype: object

In [114]:
df.dtypes

Street        object
Avg_Price    float64
Latitude     float64
Longitude    float64
dtype: object

In [115]:
# add clustering labels
GROUPED_LONDON_clustering['Cluster Labels'] = kmeans.labels_

# merge london_grouped with london_data to add latitude/longitude for each neighborhood
GROUPED_LONDON_clustering = GROUPED_LONDON_clustering.join(venues_sorted.set_index('Street'), on='Street')

GROUPED_LONDON_clustering.head(30) # check the last columns!

Street     Avg_Price   Latitude   Longitude  \
146         AIREDALE AVENUE  2.022500e+06  53.436368   -1.106305   
197           ALBION STREET  2.096667e+06 -33.819759  151.010273   
412             ANTHONY WAY  2.000000e+06  44.094651 -123.151242   
554   ASHCHURCH PARK VILLAS  2.150000e+06  51.500051   -0.242173   
673             AVENUE ROAD  2.143471e+06  51.406797   -0.049519   
762         BALLINGDON ROAD  2.105000e+06  51.454189   -0.158856   
1092       BELSIZE CRESCENT  2.000000e+06  51.549568   -0.172733   
1135      BERESFORD TERRACE  2.100000e+06  55.975743   -3.203197   
1193         BETTRIDGE ROAD  2.025000e+06  57.032038   -2.147699   
1384           BLYTHS WHARF  2.000000e+06  51.509001   -0.034583   
1435           BOSTON PLACE  2.167500e+06  51.507373   -0.340338   
1515    BRACKENBURY GARDENS  2.150000e+06  51.500623   -0.230729   
1559        BRAMSHOT AVENUE  2.177900e+06  44.606068  -75.690486   
1704        BROADLANDS ROAD  2.043000e+06  52.750115    1.566247   
1751       BRONDESBURY PARK  2.090357e+06  51.540514   -0.210336   
1775        BROOKFIELD PARK  2.150000e+06  36.931536  -76.232717   
1823         BROWNING CLOSE  2.160000e+06  52.664029    1.253144   
1867       BRYANSTON SQUARE  2.197583e+06  51.517067   -0.160365   
1911           BUNKERS HILL  2.000000e+06  53.018058   -2.380268   
2019         BYWATER STREET  2.000000e+06  42.589268  -83.443243   
2072          CALTON AVENUE  2.016667e+06  53.508760   -2.283024   
2153       CANFIELD GARDENS  2.188333e+06  51.546792   -0.179550   
2226          CARLISLE ROAD  2.200000e+06  43.399438  -79.973215   
2242          CARLYLE CLOSE  2.175000e+06  52.085813    1.135452   
2697     CHOLMELEY CRESCENT  2.025000e+06  51.572397   -0.145105   
2789      CLARE LAWN AVENUE  2.083000e+06  51.457853   -0.262927   
2857       CLEVELAND SQUARE  2.150000e+06  53.401632   -2.985580   
2989         COLINETTE ROAD  2.124375e+06  51.461939   -0.229873   
2997          COLLEGE CROSS  2.185000e+06  51.949009   -8.549684   
3042         COLVILLE PLACE  2.095000e+06  57.155024   -2.092671   

      Cluster Labels       1st Most Common Venue       2nd Most Common Venue  \
146                0          Italian Restaurant                    Pharmacy   
197                2           Indian Restaurant                        Café   
412                0                         NaN                         NaN   
554                1               Grocery Store                        Park   
673                1            Tapas Restaurant               Grocery Store   
762                2                         Pub          Italian Restaurant   
1092               0          Italian Restaurant               Movie Theater   
1135               2                         Bar              Cosmetics Shop   
1193               0                         Bar               Grocery Store   
1384               0                         Pub        Gym / Fitness Center   
1435               1                         Pub                        Café   
1515               1               Grocery Store                 Coffee Shop   
1559               3                Dance Studio                         Spa   
1704               4             Harbor / Marina             Bed & Breakfast   
1751               2                        Park         Japanese Restaurant   
1775               1              Discount Store           Recreation Center   
1823               1  Construction & Landscaping          Seafood Restaurant   
1867               3                       Hotel              Sandwich Place   
1911               0                         NaN                         NaN   
2019               0                 Pizza Place                    Pharmacy   
2072               0              Breakfast Spot           Convenience Store   
2153               3                        Café                 Coffee Shop   
2226               3                         NaN                        

In [116]:
# Create Map

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(GROUPED_LONDON_clustering['Latitude'], GROUPED_LONDON_clustering['Longitude'], GROUPED_LONDON_clustering['Street'], GROUPED_LONDON_clustering['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [117]:
GROUPED_LONDON_clustering.loc[GROUPED_LONDON_clustering['Cluster Labels'] == 0, GROUPED_LONDON_clustering.columns[[1] + list(range(5, GROUPED_LONDON_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
146   2022500.0    Italian Restaurant              Pharmacy   
412   2000000.0                   NaN                   NaN   
1092  2000000.0    Italian Restaurant         Movie Theater   
1193  2025000.0                   Bar         Grocery Store   
1384  2000000.0                   Pub  Gym / Fitness Center   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
146           Antique Shop                   Pub                   Zoo   
412                    NaN                   NaN                   NaN   
1092                Bakery            Bagel Shop      Greek Restaurant   
1193                  Park    Chinese Restaurant           Dry Cleaner   
1384    Italian Restaurant             Gastropub           Beer Garden   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
146      Electronics Store    English Restaurant  Ethiopian Restaurant   
412                    NaN                   NaN                   NaN   
1092                  Café                   Pub                 Hotel   
1193     Electronics Store    English Restaurant  Ethiopian Restaurant   
1384            Restaurant                 Trail    Turkish Restaurant   

     9th Most Common Venue 10th Most Common Venue  
146            Event Space                Exhibit  
412                    NaN                    NaN  
1092       Bed & Breakfast      Indian Restaurant  
1193           Event Space                Exhibit  
1384                  Park         Breakfast Spot

In [118]:
GROUPED_LONDON_clustering.loc[GROUPED_LONDON_clustering['Cluster Labels'] == 1, GROUPED_LONDON_clustering.columns[[1] + list(range(5, GROUPED_LONDON_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
554   2.150000e+06         Grocery Store                  Park   
673   2.143471e+06      Tapas Restaurant         Grocery Store   
1435  2.167500e+06                   Pub                  Café   
1515  2.150000e+06         Grocery Store           Coffee Shop   
1775  2.150000e+06        Discount Store     Recreation Center   

     3rd Most Common Venue     4th Most Common Venue  5th Most Common Venue  \
554                    Pub  Mediterranean Restaurant            Coffee Shop   
673                   Park              Tram Station                    Zoo   
1435           Coffee Shop               Pizza Place            Supermarket   
1515                   Pub                     Hotel  Street Food Gathering   
1775          Home Service         Electronics Store     English Restaurant   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
554      Fish & Chips Shop     Indian Restaurant                Bakery   
673             Donut Shop     Electronics Store    English Restaurant   
1435                  Park                   Gym              Bus Stop   
1515                   Gym    Falafel Restaurant    Chinese Restaurant   
1775  Ethiopian Restaurant           Event Space               Exhibit   

          9th Most Common Venue 10th Most Common Venue  
554         Japanese Restaurant                   Café  
673        Ethiopian Restaurant            Event Space  
1435          Fish & Chips Shop             Canal Lock  
1515  Latin American Restaurant      Indian Restaurant  
1775                 Eye Doctor     Falafel Restaurant

In [119]:
GROUPED_LONDON_clustering.loc[GROUPED_LONDON_clustering['Cluster Labels'] == 2, GROUPED_LONDON_clustering.columns[[1] + list(range(5, GROUPED_LONDON_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
197   2.096667e+06     Indian Restaurant                  Café   
762   2.105000e+06                   Pub    Italian Restaurant   
1135  2.100000e+06                   Bar        Cosmetics Shop   
1751  2.090357e+06                  Park   Japanese Restaurant   
2989  2.124375e+06     Convenience Store           Coffee Shop   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
197                    Gym            Sports Bar           Coffee Shop   
762          Women's Store             Wine Shop                Bakery   
1135                   Pub            Bike Trail           Supermarket   
1751           Coffee Shop                   Pub        Farmers Market   
2989        Cricket Ground                  Café             Gastropub   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
197           Dessert Shop                   Pub                  Park   
762            Coffee Shop          Antique Shop                  Café   
1135        Farmers Market    English Restaurant  Ethiopian Restaurant   
1751         Deli / Bodega  Gym / Fitness Center                   Zoo   
2989         Grocery Store          Tennis Court              Bus Stop   

     9th Most Common Venue 10th Most Common Venue  
197                    Zoo     English Restaurant  
762                    Zoo     English Restaurant  
1135           Event Space                Exhibit  
1751     Electronics Store     English Restaurant  
2989  Gym / Fitness Center                    Zoo

In [120]:
GROUPED_LONDON_clustering.loc[GROUPED_LONDON_clustering['Cluster Labels'] == 3, GROUPED_LONDON_clustering.columns[[1] + list(range(5, GROUPED_LONDON_clustering.shape[1]))]].head()

Avg_Price       1st Most Common Venue 2nd Most Common Venue  \
1559  2.177900e+06                Dance Studio                   Spa   
1867  2.197583e+06                       Hotel        Sandwich Place   
2153  2.188333e+06                        Café           Coffee Shop   
2226  2.200000e+06                         NaN                   NaN   
2242  2.175000e+06  Construction & Landscaping                   Zoo   

     3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
1559             Bookstore                   Gym             Farmers Market   
1867           Coffee Shop   Lebanese Restaurant  Middle Eastern Restaurant   
2153    Italian Restaurant           Pizza Place             Sandwich Place   
2226                   NaN                   NaN                        NaN   
2242        Farmers Market    English Restaurant       Ethiopian Restaurant   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1559  Ethiopian Restaurant           Event Space               Exhibit   
1867                   Pub                Bakery             Hotel Bar   
2153                 Hotel         Grocery Store              Bus Stop   
2226                   NaN                   NaN                   NaN   
2242           Event Space               Exhibit            Eye Doctor   

     9th Most Common Venue 10th Most Common Venue  
1559            Eye Doctor     Falafel Restaurant  
1867    Italian Restaurant     Persian Restaurant  
2153                Bakery              Bookstore  
2226                   NaN                    NaN  
2242    Falafel Restaurant   Fast Food Restaurant

In [121]:
GROUPED_LONDON_clustering.loc[GROUPED_LONDON_clustering['Cluster Labels'] == 4, GROUPED_LONDON_clustering.columns[[1] + list(range(5, GROUPED_LONDON_clustering.shape[1]))]].head()

Avg_Price 1st Most Common Venue 2nd Most Common Venue  \
1704  2043000.0       Harbor / Marina       Bed & Breakfast   
2789  2083000.0           Golf Course      Business Service   
3286  2065000.0                   NaN                   NaN   
3762  2075000.0   Arts & Crafts Store                   Pub   
4436  2078000.0             Hotel Bar                   Pub   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1704         Boat or Ferry                   Pub        Farmers Market   
2789                  Park          Soccer Field            Eye Doctor   
3286                   NaN                   NaN                   NaN   
3762                  Park    Falafel Restaurant     Electronics Store   
4436                 Hotel  Gym / Fitness Center      Department Store   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1704    English Restaurant  Ethiopian Restaurant           Event Space   
2789     Electronics Store    English Restaurant  Ethiopian Restaurant   
3286                   NaN                   NaN                   NaN   
3762    English Restaurant  Ethiopian Restaurant           Event Space   
4436     Martial Arts Dojo     French Restaurant    Italian Restaurant   

     9th Most Common Venue 10th Most Common Venue  
1704               Exhibit             Eye Doctor  
2789           Event Space                Exhibit  
3286                   NaN                    NaN  
3762               Exhibit             Eye Doctor  
4436         Grocery Store              Multiplex

## Results and Discussion

Even though, the UK has left the European Union it is still a place that both citizens and non-citizens would like to invest especially in London.

THE RESULTS 

First of all, West London (Kensington, Chelsea, Marylebone areas) and North-West London (Camden Town, Hampsted areas) were expected to pop as a profitable investment due to the wide range of amenities and essential facilitiesand venues. Yet South-West London (Brixton and Wandsworth, areas) and also Limehouse in East London are also arising as next future venues with a wide range of amenities and facilities. Therefore, South-West London and Limehouse could be a good investment opportunity for those who do not wish to invest in luxurious neighbourhoods like Kensington.

Secondly, by using the clusters the results can be analysed. In a city like London, there will be lots of pubs and restaurants but the clusters show different patterns. As you can see, the 3rd and 4th clusters tell us that those areas are more for people who are more active as it has an harbour, golf course, dance studios and fitness centres. The clusters 0, 1 and 2 can be interesting for investors who love going to pubs cafes and restaurants.

## Conclusion

The UK withdrawal from the European Union had affected the British housing market both for the  citizens and non-citizen buyers as the government made it more expensive for landlords and overseas buyers to purchase houses. Especially, real estate in London comes at quite a steep price. 
Therefore I tried to find the wisest investment options for people who seek a house in London now that Brexit is finalised. Recommendations of profitable venues according to amenities and essential facilities such as schools, hospitals & supermarkets was also a part of this project.

Data on London properties and the relative price paid data were extracted from the HM Land Registry. For data visialisation, FourSquare API interface is used. Then the collected data from HM Land Registry and Foursquare API is merged and showed the most profitable investments in London.

The Methodology for this project will start with 1. Collection of Data then continue with 2. Exploring and Understanding Data, 3. Data Preparation and Preprocessing and 4. Modelling. I will use the k-means clustering technique will be used for this project as it is accurate, efficient and highly flexible.

In conclusion, even though, the UK has left the European Union London is still a place that both people would like to invest. Although West London and North-West London were expected to pop as a profitable investment due to the wide range of amenities and essential facilitiesand venues. Yet South-West London (Brixton and Wandsworth, areas) and also Limehouse in East London are also arising as next future venues with a wide range of amenities and facilities. Therefore, South-West London and Limehouse could be a good investment opportunity for those who do not wish to invest in luxurious neighbourhoods like Kensington.
Secondly, by using the clusters the results can be analysed. In a city like London, there will be lots of pubs and restaurants but the clusters show different patterns. As you can see, the 3rd and 4th clusters tell us that those areas are more for people who are more active as it has an harbour, golf course, dance studios and fitness centres. The clusters 0, 1 and 2 can be interesting for investors who love going to pubs cafes and restaurants.